In [2]:
import pandas as pd

In [3]:
patients=pd.read_csv('patients.csv')
treatment=pd.read_csv('treatments.csv')
adverse=pd.read_csv('adverse_reactions.csv')

In [4]:
print(patients.shape)
print(treatment.shape)
print(adverse.shape)

(503, 14)
(280, 7)
(34, 3)


In [5]:
patients

patient_id assigned_sex given_name      surname  \
0             1       female        Zoe      Wellish   
1             2       female     Pamela         Hill   
2             3         male        Jae       Debord   
3             4         male       Liêm         Phan   
4             5         male        Tim      Neudorf   
..          ...          ...        ...          ...   
498         499         male    Mustafa    Lindström   
499         500         male      Ruman      Bisliev   
500         501       female      Jinke    de Keizer   
501         502       female    Chidalu  Onyekaozulu   
502         503         male        Pat      Gersten   

                       address               city       state  zip_code  \
0         576 Brown Bear Drive  Rancho California  California   92390.0   
1    2370 University Hill Road          Armstrong    Illinois   61812.0   
2        1493 Poling Farm Road               York    Nebraska   68467.0   
3          2335 Webster Street         Woodbridge          NJ    7095.0   
4         1428 Turkey Pen Lane             Dothan          AL   36303.0   
..                         ...                ...         ...       ...   
498        2530 Victoria Court       Milton Mills          ME    3852.0   
499   494 Clarksburg Park Road             Sedona          AZ   86341.0   
500          649 Nutter Street      Overland Park          MO   64110.0   
501   3652 Boone Crockett Lane            Seattle          WA   98109.0   
502          2778 North Avenue               Burr    Nebraska   68324.0   

           country                                          contact  \
0    United States             951-719-9170ZoeWellish@superrito.com   
1    United States            PamelaSHill@cuvox.de+1 (217) 569-3204   
2    United States                 402-363-6804JaeMDebord@gustr.com   
3    United States       PhanBaLiem@jourrapide.com+1 (732) 636-8246   
4    United States                  334-515-7487TimNeudorf@cuvox.de   
..             ...                                              ...   
498  United States      207-477-0579MustafaLindstrom@jourrapide.com   
499  United States               928-284-4492RumanBisliev@gustr.com   
500  United States            816-223-6007JinkedeKeizer@teleworm.us   
501  United States  ChidaluOnyekaozulu@jourrapide.com1 360 443 2060   
502  United States             PatrickGersten@rhyta.com402-848-4923   

     birthdate  weight  height   bmi  
0    7/10/1976   121.7      66  19.6  
1     4/3/1967   118.8      66  19.2  
2    2/19/1980   177.8      71  24.8  
3    7/26/1951   220.9      70  31.7  
4    2/18/1928   192.3      27  26.1  
..         ...     ...     ...   ...  
498  4/10/1959   181.1      72  24.6  
499  3/26/1948   239.6      70  34.4  
500  1/13/1971   171.2      67  26.8  
501  2/13/1952   176.9      67  27.7  
502   5/3/1954   138.2      71  19.3  

[503 rows x 14 columns]

In [6]:
treatment

given_name       surname    auralin    novodra  hba1c_start  hba1c_end  \
0     veronika      jindrová  41u - 48u          -         7.63       7.20   
1       elliot    richardson          -  40u - 45u         7.56       7.09   
2     yukitaka      takenaka          -  39u - 36u         7.68       7.25   
3         skye    gormanston  33u - 36u          -         7.97       7.62   
4       alissa        montez          -  33u - 29u         7.78       7.46   
..         ...           ...        ...        ...          ...        ...   
275     albina      zetticci  45u - 51u          -         7.93       7.73   
276       john   teichelmann          -  49u - 49u         7.90       7.58   
277     mathea       lillebø  23u - 36u          -         9.04       8.67   
278     vallie        prince  31u - 38u          -         7.64       7.28   
279     samúel  guðbrandsson  53u - 56u          -         8.00       7.64   

     hba1c_change  
0             NaN  
1            0.97  
2             NaN  
3            0.35  
4            0.32  
..            ...  
275          0.20  
276           NaN  
277          0.37  
278          0.36  
279          0.36  

[280 rows x 7 columns]

In [7]:
adverse.head()

given_name     surname           adverse_reaction
0      berta  napolitani  injection site discomfort
1       lena        baer               hypoglycemia
2     joseph         day               hypoglycemia
3     flavia  fiorentino                      cough
4    manouck     wubbels          throat irritation

In [8]:
patients.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 503 entries, 0 to 502
Data columns (total 14 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   patient_id    503 non-null    int64  
 1   assigned_sex  503 non-null    object 
 2   given_name    503 non-null    object 
 3   surname       503 non-null    object 
 4   address       491 non-null    object 
 5   city          491 non-null    object 
 6   state         491 non-null    object 
 7   zip_code      491 non-null    float64
 8   country       491 non-null    object 
 9   contact       491 non-null    object 
 10  birthdate     503 non-null    object 
 11  weight        503 non-null    float64
 12  height        503 non-null    int64  
 13  bmi           503 non-null    float64
dtypes: float64(3), int64(2), object(9)
memory usage: 55.1+ KB


In [9]:
treatment.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 280 entries, 0 to 279
Data columns (total 7 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   given_name    280 non-null    object 
 1   surname       280 non-null    object 
 2   auralin       280 non-null    object 
 3   novodra       280 non-null    object 
 4   hba1c_start   280 non-null    float64
 5   hba1c_end     280 non-null    float64
 6   hba1c_change  171 non-null    float64
dtypes: float64(3), object(4)
memory usage: 15.4+ KB


In [10]:
adverse.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 34 entries, 0 to 33
Data columns (total 3 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   given_name        34 non-null     object
 1   surname           34 non-null     object
 2   adverse_reaction  34 non-null     object
dtypes: object(3)
memory usage: 948.0+ bytes


In [11]:
patients[patients['address'].isnull()]

patient_id assigned_sex given_name      surname address city state  \
209         210       female     Lalita  Eldarkhanov     NaN  NaN   NaN   
219         220         male         Mỹ        Quynh     NaN  NaN   NaN   
230         231       female  Elisabeth      Knudsen     NaN  NaN   NaN   
234         235       female    Martina    Tománková     NaN  NaN   NaN   
242         243         male       John      O'Brian     NaN  NaN   NaN   
249         250         male   Benjamin       Mehler     NaN  NaN   NaN   
257         258         male        Jin         Kung     NaN  NaN   NaN   
264         265       female   Wafiyyah       Asfour     NaN  NaN   NaN   
269         270       female     Flavia   Fiorentino     NaN  NaN   NaN   
278         279       female   Generosa        Cabán     NaN  NaN   NaN   
286         287         male      Lewis         Webb     NaN  NaN   NaN   
296         297       female        Chỉ          Lâm     NaN  NaN   NaN   

     zip_code country contact   birthdate  weight  height   bmi  
209       NaN     NaN     NaN   8/14/1950   143.4      62  26.2  
219       NaN     NaN     NaN    4/9/1978   237.8      69  35.1  
230       NaN     NaN     NaN   9/23/1976   165.9      63  29.4  
234       NaN     NaN     NaN    4/7/1936   199.5      65  33.2  
242       NaN     NaN     NaN   2/25/1957   205.3      74  26.4  
249       NaN     NaN     NaN  10/30/1951   146.5      69  21.6  
257       NaN     NaN     NaN   5/17/1995   231.7      69  34.2  
264       NaN     NaN     NaN   11/3/1989   158.6      63  28.1  
269       NaN     NaN     NaN   10/9/1937   175.2      61  33.1  
278       NaN     NaN     NaN  12/16/1962   124.3      69  18.4  
286       NaN     NaN     NaN    4/1/1979   155.3      68  23.6  
296       NaN     NaN     NaN   5/14/1990   181.1      63  32.1

In [12]:
patients.describe()

patient_id      zip_code      weight      height         bmi
count  503.000000    491.000000  503.000000  503.000000  503.000000
mean   252.000000  49084.118126  173.434990   66.634195   27.483897
std    145.347859  30265.807442   33.916741    4.411297    5.276438
min      1.000000   1002.000000   48.800000   27.000000   17.100000
25%    126.500000  21920.500000  149.300000   63.000000   23.300000
50%    252.000000  48057.000000  175.300000   67.000000   27.200000
75%    377.500000  75679.000000  199.500000   70.000000   31.750000
max    503.000000  99701.000000  255.900000   79.000000   37.700000

In [13]:
treatment.describe()

hba1c_start   hba1c_end  hba1c_change
count   280.000000  280.000000    171.000000
mean      7.985929    7.589286      0.546023
std       0.568638    0.569672      0.279555
min       7.500000    7.010000      0.200000
25%       7.660000    7.270000      0.340000
50%       7.800000    7.420000      0.380000
75%       7.970000    7.570000      0.920000
max       9.950000    9.580000      0.990000

In [14]:
patients.sample(5)

patient_id assigned_sex given_name   surname              address  \
389         390         male      Yegor  Uspensky     4900 Philli Lane   
86           87         male   Philemon   Abdulov     1341 Chatham Way   
224         225       female   Veronika  Jindrová       3641 Selah Way   
491         492       female    Vanessa  Ferguson  241 Freshour Circle   
146         147         male     Haylom     Nebay   1736 Parrill Court   

             city state  zip_code        country  \
389         Tulsa    OK   74133.0  United States   
86   Temple Hills    MD   20031.0  United States   
224   Brattleboro    VT    5301.0  United States   
491   San Antonio    TX   78205.0  United States   
146  East Chicago    IN   46312.0  United States   

                                         contact  birthdate  weight  height  \
389        918-459-9811YegorUspensky@fleckens.hu  9/10/1956   136.8      65   
86         240-695-5212PhilemonAbdulov@rhyta.com   7/7/1943   171.4      70   
224  802-614-0812VeronikaJindrova@jourrapide.com  1/15/1981   171.4      62   
491   210-222-8684VanessaFerguson@jourrapide.com  9/21/1950   149.8      67   
146          HaylomNebay@gustr.com1 219 391 7729   9/1/1963   185.9      70   

      bmi  
389  22.8  
86   24.6  
224  31.3  
491  23.5  
146  26.7

In [15]:
patients[patients['address'].duplicated()]

patient_id assigned_sex given_name     surname             address  \
29           30         male       Jake    Jakobsen   648 Old Dear Lane   
219         220         male         Mỹ       Quynh                 NaN   
229         230         male       John         Doe     123 Main Street   
230         231       female  Elisabeth     Knudsen                 NaN   
234         235       female    Martina   Tománková                 NaN   
237         238         male       John         Doe     123 Main Street   
242         243         male       John     O'Brian                 NaN   
244         245         male       John         Doe     123 Main Street   
249         250         male   Benjamin      Mehler                 NaN   
251         252         male       John         Doe     123 Main Street   
257         258         male        Jin        Kung                 NaN   
264         265       female   Wafiyyah      Asfour                 NaN   
269         270       female     Flavia  Fiorentino                 NaN   
277         278         male       John         Doe     123 Main Street   
278         279       female   Generosa       Cabán                 NaN   
282         283       female      Sandy      Taylor  2476 Fulton Street   
286         287         male      Lewis        Webb                 NaN   
296         297       female        Chỉ         Lâm                 NaN   
502         503         male        Pat     Gersten   2778 North Avenue   

            city     state  zip_code        country  \
29   Port Jervis  New York   12771.0  United States   
219          NaN       NaN       NaN            NaN   
229     New York        NY   12345.0  United States   
230          NaN       NaN       NaN            NaN   
234          NaN       NaN       NaN            NaN   
237     New York        NY   12345.0  United States   
242          NaN       NaN       NaN            NaN   
244     New York        NY   12345.0  United States   
249          NaN       NaN       NaN            NaN   
251     New York        NY   12345.0  United States   
257          NaN       NaN       NaN            NaN   
264          NaN       NaN       NaN            NaN   
269          NaN       NaN       NaN            NaN   
277     New York        NY   12345.0  United States   
278          NaN       NaN       NaN            NaN   
282     Rainelle        WV   25962.0  United States   
286          NaN       NaN       NaN            NaN   
296          NaN       NaN       NaN            NaN   
502         Burr  Nebraska   68324.0  United States   

                                        contact   birthdate  weight  height  \
29   JakobCJakobsen@einrot.com+1 (845) 858-7707    8/1/1985   155.8      67   
219                                         NaN    4/9/1978   237.8      69   
229                 johndoe@email.com1234567890    1/1/1975   180.0      72   
230                                         NaN   9/23/1976   165.9      63   
234                                         NaN    4/7/1936   199.5      65   
237                 johndoe@email.com1234567890    1/1/1975   180.0      72   
242                                         NaN   2/25/1957   205.3      74   
244                 johndoe@email.com1234567890    1/1/1975   180.0      72   
249                                         NaN  10/30/1951   146.5      69   
251                 johndoe@email.com1234567890    1/1/1975   180.0      72   
257                                         NaN   5/17/1995   231.7      69   
264                                         NaN   11/3/1989   158.6      63   
269                                         NaN   10/9/1937   175.2      61   
277                 johndoe@email.com1234567890    1/1/1975   180.0      72   
278                                         NaN  12/16/1962   124.3      69   
282        304-438-2648SandraCTaylor@dayrep.com  10/23/1960   206.1      64   
286                                         NaN    4/1/1979 

Quality
[Completeness, Validity, Accuracy, Consistency]

Patients
Zip code col is stored in decimal format[V]
Certain zip codes of 4 digits[V]
For the state col sometimes full names are use and other times short forms are used[IC]
Missing values for contact details of 12 patients[C]
Incorrect data types for assigned_sex and birthdate cols[?]
Erroneous data for weights and heights[Accuracy]
Inconsistent way of storing phone numbers in contact col[IC]
Duplicate and not useful data in dataset[IC]
         
Treatments
hba1c_change col has missing values[C]
Novodra and Auralin cols has u attached to numbers[V]
Only 280 rows are there, instead of 350[C]
Name and surname col is in lower case[IC]
Incorrect data type for auralin and novadra cols[V]
hba1c_change 9 instead of 4[Accuracy]
use of dashes instead of Nan in aurolin and novodra cols[IC]
         
Adverse Reactions
Name and surname col is in lower case[IC]
         
Tidiness
Patients
Phone and email together in one col i.e. contact
         
Treatments
3 variables stored in auralin and novodra cols
         
Adverse Reactions
No point of using this table

Cleaning Data
You should follow the order provided below:

Completeness Issues
Tidiness Issues
Remaining Data quality issues like validity, accuracy and consistency

Steps involved in Data cleaning
Define
Code
Test

Always make sure to create a copy of your pandas dataframe before you start the cleaning process

In [16]:
# Creating copies

patients_copy=patients.copy()
treatment_copy=treatment.copy()
adverse_copy=adverse.copy()

In [18]:
patients_copy[patients_copy['address'].isnull()]

patient_id assigned_sex given_name      surname address city state  \
209         210       female     Lalita  Eldarkhanov     NaN  NaN   NaN   
219         220         male         Mỹ        Quynh     NaN  NaN   NaN   
230         231       female  Elisabeth      Knudsen     NaN  NaN   NaN   
234         235       female    Martina    Tománková     NaN  NaN   NaN   
242         243         male       John      O'Brian     NaN  NaN   NaN   
249         250         male   Benjamin       Mehler     NaN  NaN   NaN   
257         258         male        Jin         Kung     NaN  NaN   NaN   
264         265       female   Wafiyyah       Asfour     NaN  NaN   NaN   
269         270       female     Flavia   Fiorentino     NaN  NaN   NaN   
278         279       female   Generosa        Cabán     NaN  NaN   NaN   
286         287         male      Lewis         Webb     NaN  NaN   NaN   
296         297       female        Chỉ          Lâm     NaN  NaN   NaN   

     zip_code country contact   birthdate  weight  height   bmi  
209       NaN     NaN     NaN   8/14/1950   143.4      62  26.2  
219       NaN     NaN     NaN    4/9/1978   237.8      69  35.1  
230       NaN     NaN     NaN   9/23/1976   165.9      63  29.4  
234       NaN     NaN     NaN    4/7/1936   199.5      65  33.2  
242       NaN     NaN     NaN   2/25/1957   205.3      74  26.4  
249       NaN     NaN     NaN  10/30/1951   146.5      69  21.6  
257       NaN     NaN     NaN   5/17/1995   231.7      69  34.2  
264       NaN     NaN     NaN   11/3/1989   158.6      63  28.1  
269       NaN     NaN     NaN   10/9/1937   175.2      61  33.1  
278       NaN     NaN     NaN  12/16/1962   124.3      69  18.4  
286       NaN     NaN     NaN    4/1/1979   155.3      68  23.6  
296       NaN     NaN     NaN   5/14/1990   181.1      63  32.1

a. Completeness Issues
Missing values for contact details of 12 patients
Only 280 rows are there, instead of 350
hba1c_change col has missing values

Define

In [19]:
# Subtract hba1c_start from hba1c_end to get hba1c_change

Code

In [20]:
treatment_copy['hba1c_change']=treatment_copy['hba1c_start']-treatment_copy['hba1c_end']

Test

In [21]:
treatment_copy.info()

treatment_copy.describe()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 280 entries, 0 to 279
Data columns (total 7 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   given_name    280 non-null    object 
 1   surname       280 non-null    object 
 2   auralin       280 non-null    object 
 3   novodra       280 non-null    object 
 4   hba1c_start   280 non-null    float64
 5   hba1c_end     280 non-null    float64
 6   hba1c_change  280 non-null    float64
dtypes: float64(3), object(4)
memory usage: 15.4+ KB


hba1c_start   hba1c_end  hba1c_change
count   280.000000  280.000000    280.000000
mean      7.985929    7.589286      0.396643
std       0.568638    0.569672      0.060168
min       7.500000    7.010000      0.200000
25%       7.660000    7.270000      0.350000
50%       7.800000    7.420000      0.390000
75%       7.970000    7.570000      0.440000
max       9.950000    9.580000      0.540000

b. Tidiness Issues
Phone + email in contact col patients
Novodra and Aurolin col treatment
Merge the adverse table to treatment table

Define

In [22]:
# Novodra and Aurolin using the melt function

Code

In [23]:
treatment_copy=pd.melt(treatment_copy, id_vars=['given_name', 'surname', 'hba1c_start', 'hba1c_end', 'hba1c_change'],
                           var_name='treatment', value_name='dose')

In [24]:
treatment_copy=treatment_copy[treatment_copy['dose']!='-']

In [ ]:
treatment_copy['hba1c_start'],treatment_copy['hba1c_end']=treatment_copy['dose'].str.split('-').str

In [32]:
treatment_copy.drop(columns={'dose'}, inplace=True)

Test

In [33]:
treatment_copy.head()

given_name     surname  hba1c_start  hba1c_end  hba1c_change treatment
0   veronika    jindrová         7.63       7.20          0.43   auralin
3       skye  gormanston         7.97       7.62          0.35   auralin
6     sophia      haugen         7.65       7.27          0.38   auralin
7      eddie      archer         7.89       7.55          0.34   auralin
9       asia     woźniak         7.76       7.37          0.39   auralin

What is Assessing Data?
The data analysis process consists of the following steps

Asking Questions
Data Wrangling

a. Gathering Data i. CSV files ii. APIs iii. Web Scraping iv. Databases b. Assessing Data
c. Cleaning Data

Exploratory Data Analysis
Drawing Conclusion
Comunicating Results

Assessing Data
In this step, the data is to be understood more deeply. Before implementing methods to clean it, you will definitely need to have a better idea about what the data is about.

Types of Unclean Data
There are 2 kinds of unclean data

Dirty Data (Data with Quality issues):
Dirty data, also known as low quality data. Low quality data has content issues.
Messy Data (Data with tidiness issues)
Messy data, also known as untidy data. Untidy data has structural issues.
1. Dirty Data
When the data has following kinds of issues:

Duplicated data
Missing Data
Corrupt Data
Inaccurate Data
This kind of data is known as Dirty data

2. Messy Data
When the data has structural issue is it known as Messy or untidy data.

Tidy data has the following properties:

Each variable forms a column
Each observation forms a row
Each observational unit forms a table
If a data violates any one of the above it is known as untidy data

Example:

Types of Assessment
There are 2 types of assessment styles

Manual - Looking through the data manually in google sheets
Programmatic - By using pandas functions such as info(), describe() or sample()
Steps in Assessment
There are 2 steps involved in Assessment

Discover
Document
Data Quality Dimensions
Completeness Issues
Validity Issues
Accuracy Issues
Consistency Issues